baybe/utils/interval.py

```diff
@@ -41,40 +45,60 @@ class Interval:
     """The upper end of the interval."""
 
     @upper.validator
-    def _validate_order(self, _: Any, value: float):  # noqa: DOC101, DOC103
+    def _validate_order(self, _: Any, upper: float) -> None:  # noqa: DOC101, DOC103
         """Validate the order of the interval bounds.
 
         Raises:
             ValueError: If the upper end is not larger than the lower end.
         """
-        if value <= self.lower:
+        if upper <= self.lower:
             raise ValueError(
-                f"The upper interval bound (provided value: {value}) must be larger "
+                f"The upper interval bound (provided value: {upper}) must be larger "
                 f"than the lower bound (provided value: {self.lower})."
             )
 
     @property
-    def is_finite(self):
-        """Check whether the interval is finite."""
-        return np.isfinite(self.lower) and np.isfinite(self.upper)
+    def is_bounded(self) -> bool:
+        """Check if the interval is bounded."""
+        return self.is_left_bounded and self.is_right_bounded
 
     @property
-    def is_bounded(self):
-        """Check whether the interval is bounded."""
-        return np.isfinite(self.lower) or np.isfinite(self.upper)
+    def is_left_bounded(self) -> bool:
+        """Check if the interval is left-bounded."""
+        return np.isfinite(self.lower)
+
+    @property
+    def is_right_bounded(self) -> bool:
+        """Check if the interval is right-bounded."""
+        return np.isfinite(self.upper)
+
+    @property
+    def is_half_bounded(self) -> bool:
+        """Check if the interval is half-bounded."""
+        return self.is_left_bounded ^ self.is_right_bounded
+
+    @property
+    def is_finite(self) -> bool:
+        """Check whether the interval is finite."""
+        warnings.warn(
+            "The use of 'Interval.is_finite' is deprecated and will be disabled in "
+            "a future version. Use 'Interval.is_bounded' instead.",
+            DeprecationWarning,
+        )
+        return np.isfinite(self.lower) and np.isfinite(self.upper)
 
     @property
-    def center(self):
-        """The center of the interval. Only applicable for finite intervals."""
-        if not self.is_finite:
+    def center(self) -> float:
+        """The center of the interval. Only applicable for bounded intervals."""
+        if not self.is_bounded:
```

Comment:
doesnt it make more sense to return `None` rather than throwing an exception?

Reply:
For me, a relatively clear "no". Because: doing so would change the return type, and this is a common antipattern for good reason: https://docs.quantifiedcode.com/python-anti-patterns/maintainability/returning_more_than_one_variable_type_from_function_call.html?highlight=exception
Basically, one reason to have a custom exception system is to avoid exactly that. 

So I will close this thread, but if you are unhappy, feel free to reopen for further discussion 🙃

Reply:
This is getting to the crux. The page speaks about should raise an error if a problem is encountered. Imo we don't have a situation like that here, we have a property that can or not exist, neither situations are `errors`. If my user asks my interval whats its center and it has no center, its quite literally telling my user "I have no center" aka `None`. Your variant is also telling the user that, but with a hammer to the face (aka exception).

My variant would enable this pattern
```python
if center := interv.center():
   # beep
else:
   # boop
```
while the existing variant requires:
```python
try:
   center = interv.center()
   # beep
except InfiniteIntervalError:
   # boop
```

Not only is the first variant more `gucci` to the eye, but theres also no need to keep track, import and remember what type of exception this would raise. 

And lastly, it is not recommended to put if-else situations into except statements as they will be slow in the except part. Hence I would perhaps even argue that that is the antipattern we should avoid here (in fact we have several hacky code parts where we use `except` blocks to treat if-else kind of situations for non-error situations).

Reply:
Hi @Scienfitz
1. Appreciate sharing your thoughts
2. Like the comparison with the hammer 😄 

Upfront: we can also quickly talk about it in the next dev. Nonetheless, here what I think about your answers:
* Yes, I agree that your pattern is more `gucci` to the eye and requires less work, and yes, detailed exception handling is the price you pay for exactness, but I could argue you don't compare against the right situation to start with (and btw: the patterns is also wrong, see end of post ;D). Instead of querying the value (and being hit with the hammer), you could just as well query for the precondition that needs to hold, and in the code logic of the calling code, that  could even be the natural thing to do:
  ```python
  if interval.is_bounded:
     # custom logic involving the use of interval.center
  else:
     # boop
  ```

* Still, you can safely ignore the previous point if you want. The much bigger concern is that in your variant, it suddenly becomes the **responsibility of the caller** to **always** assert that interval `is not None` before it can apply any logic to the returned function value. This becomes necessary for two reasons
  * If you did not do it, your program would no longer be type correct. Look at the actual code where we use it: `lookup[target.name] - target.bounds.center`
    Without the `if interval.center is not None` check upfront, this line could potentially evaluate to "some number minus None". All linters will rightfully complain!
  * Even if you ignored the linters, you will get completely useless error messages like `TypeError: unsupported operand type(s) for -: 'int' and 'NoneType'` in the worst case.
 
Lastly: Note that your pattern is actually incorrect, that is, you really HAVE TO write `if interval.center is not None` instead of `if center := interv.center()` because your expression would evaluate to `False` if the interval happens to be 0.0 (which is a rather common case!)

Let's continue in the meeting :D

Reply:
1) of course it's the responsibility of the caller to distinguish `None` or `float`. After all, these are both perfectly valid returns of `inverval.center` in my view. **I think this is the core of the disagreement**. If I ask an interval for its center and is has no center, there is neither a fault of the caller nor the interval. It simply has no center and thats valid. No error has been encountered and so no exception should be raised.
If I ask a human on the street `how is your brother?` and they have no brother, what is your preferred answer: i) human says `I have no brother.` or is it ii) human rolls on the floor and yells `THIS IS AN INVALID QUESTION. YOU SHOULD HAVE PERFORMED A BACKGROUND CHECK FIRST`.

2) Ok my pattern suggestion was bugged but quite simply fixed by adding `is not None`. It sill looks better than the alternatives. The variant with `is_bounded` works in both our variants, however i) you dont have any way of enforcing the user to use this instead of center. Implicitly assuming that a user has to check `is_bounded` before they try and access `.center` is really really bad and ii) theres no clear name connection between `is_bounded` and `center` and you can only ensure that a reader gets it by adding an explicit comment -> less optimal. 
Instead `if center:=interv.center is not None:` or `if interv.center is not None:` quite literally reads humanlike `if the interval has a center:`

3) **Wouldn't be a linter complaining about incorrect type in my variant is actually the best outcome?** as it enforced the coder to deal with all possible situations (instead of not dealing with the problem and declaring the solution be a hammer with a nicely engraved human-readable message is thrown at the users face)

Reply:
@Scienfitz: Changed as agreed in the meeting. Have a look and resolve if OK (+ accept PR if there are no other issues).

Reply:
the fix in `simulation` is not here, assume you wanted to change it in the mypy overhaul of `simulation`?

Reply:
yes, exactly. right now, no fix needed since `simulation` is not yet type checked